sys/senml/senml.c

```diff
@@ -0,0 +1,325 @@
+            input[t->end] = '\0';
+
+            if (array_num == 0) {
+                if (!strcmp(tmp_key, SJ_VERSION)) {
+                    pack->base_info->version = (uint8_t)atoi(tmp_val);
+                }
+                else if (!strcmp(tmp_key, SJ_BASE_NAME)) {
+                    pack->base_info->base_name = tmp_val;
+                }
+                else if (!strcmp(tmp_key, SJ_BASE_TIME)) {
+                    pack->base_info->base_time = atof(tmp_val);
+                }
+                else if (!strcmp(tmp_key, SJ_BASE_UNIT)) {
+                    pack->base_info->base_unit = tmp_val;
+                }
+                else if (!strcmp(tmp_key, SJ_BASE_VALUE)) {
+                    pack->base_info->base_value = atof(tmp_val);
+                }
+            }
+
+            if (!strcmp(tmp_key, SJ_NAME)) {
+                (pack->records)[array_num].name = tmp_val;
+            }
+            else if (!strcmp(tmp_key, SJ_UNIT)) {
+                (pack->records)[array_num].unit = tmp_val;
+            }
+            else if (!strcmp(tmp_key, SJ_VALUE)) {
+                (pack->records)[array_num].value_type = SENML_TYPE_FLOAT;
+                (pack->records)[array_num].value.f = atof(tmp_val);
+            }
+            else if (!strcmp(tmp_key, SJ_STRING_VALUE)) {
+                (pack->records)[array_num].value_type = SENML_TYPE_STRING;
+                (pack->records)[array_num].value.s = tmp_val;
+            }
+            else if (!strcmp(tmp_key, SJ_BOOL_VALUE)) {
+                (pack->records)[array_num].value_type = SENML_TYPE_BOOL;
+                if (!strcmp(tmp_val, "true")) {
+                    (pack->records)[array_num].value.b = true;
+                }
+                else if (!strcmp(tmp_val, "false")) {
+                    (pack->records)[array_num].value.b = false;
+                }
+            }
+            else if (!strcmp(tmp_key, SJ_DATA_VALUE)) {
+                (pack->records)[array_num].value_type = SENML_TYPE_BINARY;
+                (pack->records)[array_num].value.d = tmp_val;
+            }
+            else if (!strcmp(tmp_key, SJ_VALUE_SUM)) {
+                (pack->records)[array_num].value_sum = atof(tmp_val);
+            }
+            else if (!strcmp(tmp_key, SJ_TIME)) {
+                (pack->records)[array_num].time = atof(tmp_val);
+            }
+            else if (!strcmp(tmp_key, SJ_UPDATE_TIME)) {
+                (pack->records)[array_num].update_time = atof(tmp_val);
+            }
+        }
+    }
+
+    pack->num = (array_num + 1);
+
+    return 0;
+}
+
+
+int senml_encode_json_s(const senml_pack_t *pack, char *output, size_t len)
+{
+    size_t insert_pos = 0;
+
+    output[0] = '[';
+    insert_pos++;
+    len--;
+
+    int chars_written;
+
+    if (!pack->base_info) {
+        goto encode_records;
+    }
+
+    output[insert_pos] = '{';
+    insert_pos++;
+    len--;
+
+    // using len - 2 to leave space for closing } and ]
```

Comment:
As `size_t` is unsigned, `len` may underflow if `len < 2`, results in buffer overrun.